### PR TITLE
Return the list of errors that cause validity/compatibility issues

### PR DESCRIPTION
**Pull request by adiohana**

### CODE REVIEW

**Comment by adiohana** [comment]
#712

File: app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityExecutionResult.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    public Set<Difference> getIncompatibleDifferences() {
+        return incompatibleDifferences;
+    }
+
+    public void setIncompatibleDifferences(Set<Difference> incompatibleDifferences) {
+        this.incompatibleDifferences = incompatibleDifferences;
+    }
+
+    public static CompatibilityExecutionResult empty(boolean isCompatible) {
```
Comment (ronsher) [inline app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityExecutionResult.java:37]: When should someone return an empty compatibility result? What does it mean?

File: app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityChecker.java [hunk 2]
```diff
@@ -37,17 +38,17 @@
...
             throw new IllegalStateException("existingArtifacts contains null element(s)");
         }
-        return isCompatibleWith(
-            compatibilityLevel,
-            existingArtifacts.stream().map(ContentHandle::content).collect(Collectors.toList()),
-            proposedArtifact.content()
+        return testCompatibility(
+                compatibilityLevel,
+                existingArtifacts.stream().map(ContentHandle::content).collect(Collectors.toList()),
+                proposedArtifact.content()
```
Comment (ronsher) [inline app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityChecker.java:51]: Indentation isn't compatible with existing code

File: app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityExecutionResult.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import io.apicurio.registry.rules.compatibility.jsonschema.diff.Difference;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Created by aohana
+ */
+public class CompatibilityExecutionResult {
```
Comment (ronsher) [inline app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityExecutionResult.java:11]: Can you add to the comment some explanation on this class?

File: app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityRuleExecutor.java [hunk 4]
```diff
@@ -56,14 +61,31 @@ public void execute(RuleContext context) throws RuleViolationException {
...
-            existingArtifacts,
-            context.getUpdatedContent())
-        ) {
-            throw new RuleViolationException(String.format("Incompatible artifact: %s [%s]",
-                context.getArtifactId(), context.getArtifactType()),
-                RuleType.COMPATIBILITY, context.getConfiguration());
+        CompatibilityExecutionResult compatibilityExecutionResult = checker.testCompatibility(
+                level,
+                existingArtifacts,
+                context.getUpdatedContent());
```
Comment (ronsher) [inline app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityRuleExecutor.java:67]: indentation too deep

File: app/src/main/java/io/apicurio/registry/rules/compatibility/jsonschema/JsonSchemaDiffLibrary.java [hunk 5]
```diff
@@ -45,12 +48,12 @@ public static DiffContext findDifferences(String original, String updated) {
             JSONObject updatedJson = MAPPER.readValue(updated, JSONObject.class);
 
             Schema originalSchema = SchemaLoader.builder()
-                .schemaJson(originalJson)
-                .build().load().build();
+                    .schemaJson(originalJson)
+                    .build().load().build();
```
Comment (ronsher) [inline app/src/main/java/io/apicurio/registry/rules/compatibility/jsonschema/JsonSchemaDiffLibrary.java:52]: indentation

File: app/src/main/java/io/apicurio/registry/rules/compatibility/jsonschema/diff/DiffContext.java [hunk 6]
```diff
@@ -85,12 +86,12 @@ private void addToDifferenceSets(Difference difference) {
...
-            .pathUpdated(pathUpdated)
-            .subSchemaOriginal(Objects.toString(originalSubchema)) // make sure toString is good enough
-            .subSchemaUpdated(Objects.toString(updatedSubchema))
-            .build();
+                .diffType(type)
+                .pathOriginal("")
+                .pathUpdated(pathUpdated)
+                .subSchemaOriginal(Objects.toString(originalSubchema)) // make sure toString is good enough
+                .subSchemaUpdated(Objects.toString(updatedSubchema))
+                .build();
```
Comment (ronsher) [inline app/src/main/java/io/apicurio/registry/rules/compatibility/jsonschema/diff/DiffContext.java:94]: indentation

File: app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityExecutionResult.java (hunk 1 above)
Comment (adiohana) [reply app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityExecutionResult.java:37]: all the schema types (except JSON) currently don't have information on the compatibility error so they will return isCompatible (true/false) and an empty diff set

File: app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityExecutionResult.java [hunk 7]
```diff
@@ -7,6 +7,9 @@
 
 /**
  * Created by aohana
+ *
+ * Holds the result for compatibility check
```
Comment (ronsher) [inline app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityExecutionResult.java:11]: ...result for **a** compatibility check

**Comment by ronsher** [comment]
LGTM. @EricWittmann ?

**Comment by adiohana** [comment]
Hi @EricWittmann 
will appreciate your feedback here.
thanks !

**Comment by EricWittmann** [comment]
Sorry for the delay!  Busy busy busy...  I'm looking now.   :)

File: common/src/main/resources/META-INF/openapi.json [hunk 8]
```diff
@@ -2233,7 +2233,7 @@
                 }
             },
             "RuleViolationCause": {
-                "title": "Root Type for RuleViolationCAuse",
+                "title": "Root Type for RuleViolationCause",
```
Comment (EricWittmann) [inline common/src/main/resources/META-INF/openapi.json:2236]: :) nice catch

**Review by EricWittmann — APPROVED** [review]
Looks pretty good, thanks for all the work and for going back and forth on this with me.  I think there are some minor tweaks I may want to make, particularly with validity.  But I'll followup on that after this is merged.

**Comment by adiohana** [comment]
thanks @EricWittmann 
as you said, we still have some work here. but we can take care of it later on.

I noticed that after merging latest code from master branch there are some tests failures, I pushed a fix, it looks good now.

**Review by EricWittmann — APPROVED** [review]

**Pull request merged**